lib/include_utils.js

```diff
@@ -0,0 +1,99 @@
+module.exports.buildOneToOneIdentityMapWithOrigKeys = buildOneToOneIdentityMapWithOrigKeys;
+module.exports.buildOneToManyIdentityMapWithOrigKeys = buildOneToManyIdentityMapWithOrigKeys;
+module.exports.join = join;
+module.exports.KVMap = KVMap;
+
+/**
+ * Effectively builds associative map on id -> object relation and stores original keys.
+ * Map returned in form of object with ids in keys and object as values.
+ * @param objs array of objects to build from
+ * @param idName name of property to be used as id. Such property considered to be unique across array.
+ * In case of collisions last wins. For non-unique ids use buildOneToManyIdentityMap()
+ * @returns {} object where keys are ids and values are objects itself
+ */
+function buildOneToOneIdentityMapWithOrigKeys(objs, idName) {
+  var kvMap = new KVMap();
+  for(var i = 0; i < objs.length; i++) {
+    var obj = objs[i];
+    var id = obj[idName];
+    kvMap.set(id, obj);
+  }
+  return kvMap;
+}
+
+function buildOneToManyIdentityMapWithOrigKeys(objs, idName) {
+  var kvMap = new KVMap();
+  for(var i = 0; i < objs.length; i++) {
+    var obj = objs[i];
+    var id = obj[idName];
+    var value = kvMap.get(id) || [];
+    value.push(obj);
+    kvMap.set(id, value);
+  }
+  return kvMap;
+}
+
+
+/**
+ * Yeah, it joins. You need three things id -> obj1 map, id -> [obj2] map and merge function.
+ * This functions will take each obj1, locate all data to join in map2 and call merge function.
+ * @param oneToOneIdMap
+ * @param oneToManyIdMap
+ * @param mergeF  function(obj, objectsToMergeIn)
+ */
+function join(oneToOneIdMap, oneToManyIdMap, mergeF) {
+  var ids = oneToOneIdMap.getKeys();
+  for(var i = 0; i < ids.length; i++) {
+    var id = ids[i];
+    var obj = oneToOneIdMap.get(id);
+    var objectsToMergeIn = oneToManyIdMap.get(id) || [];
+    mergeF(obj, objectsToMergeIn);
+  }
+}
+
+
+/**
+ * Map with arbitrary keys and values. User .set() and .get() to work with values instead of []
+ * @returns {{set: Function, get: Function, remove: Function, exist: Function, getKeys: Function}}
+ * @constructor
+ */
+function KVMap(){
+  var _originalKeyFieldName = 'originalKey';
+  var _valueKeyFieldName = 'value';
+  var _dict = {};
+  var keyToString = function(key){ return key.toString() };
```

Comment:
Ok I found a way of getting the key dump. 
The problem seems to have appeared because I use the fields query.
```
"fields": [
   "position",
   "value"
]
```
So in buildOneToOneIdentityMapWithOrigKeys(objs, idName) I ended up with 
`obj = { location: {lat: 45, lng: 45}, value: 5 }`
and
`idName = 'id'`
which resulted in the key being undefined.